### PR TITLE
Add support for enum forward declaration

### PR DESCRIPTION
**Pull request by Jezurko**

### CODE REVIEW

File: test/vast/Dialect/HighLevel/enum-f.c [hunk 1]
```diff
@@ -0,0 +1,20 @@
+// RUN: vast-front -vast-emit-mlir=hl -o - %s | FileCheck %s
+
+// CHECK: hl.enum "kobj_ns_type" : !hl.int< unsigned >
```
Comment (xlauko) [inline test/vast/Dialect/HighLevel/enum-f.c:3]: But this is not forward declaration.

File: include/vast/Dialect/HighLevel/HighLevelOps.td [hunk 2]
```diff
@@ -235,10 +235,13 @@ def EnumDeclOp
...
       "mlir::Type":$type,
       "BuilderCallback":$constants
+    )>,
+    OpBuilder<(ins
+      "llvm::StringRef":$name
     )>
   ];
 
-  let assemblyFormat = [{ $name attr-dict `:` $type $constants }];
+  let assemblyFormat = [{ $name attr-dict `:` ($type^ $constants)? }];
```
Comment (xlauko) [inline include/vast/Dialect/HighLevel/HighLevelOps.td:244]: Does missing type implies missing constants?

File: include/vast/Translation/CodeGenDeclVisitor.hpp [hunk 3]
```diff
@@ -609,7 +609,36 @@ namespace vast::cg {
...
+
+                if (!decl->isComplete()) {
+                    return context().enumdecls.lookup(prev);
+                }
+
+                while (prev) {
+                    if (auto prev_op = context().enumdecls.lookup(prev)) {
+                        VAST_ASSERT(!prev->isComplete());
+                        prev_op.setType(visit(decl->getIntegerType()));
+                        InsertionGuard guard(builder());
```
Comment (xlauko) [inline include/vast/Translation/CodeGenDeclVisitor.hpp:623]: use ` auto guard = insertion_guard();` which infers correct builder

File: include/vast/Dialect/HighLevel/HighLevelOps.td (hunk 2 above)
Comment (Jezurko) [reply include/vast/Dialect/HighLevel/HighLevelOps.td:244]: yes

File: lib/vast/Dialect/HighLevel/HighLevelOps.cpp [hunk 4]
```diff
@@ -202,6 +202,11 @@ namespace vast::hl
         InsertionGuard guard(bld);
         build_region(bld, st, constants);
     }
+    void EnumDeclOp::build(Builder &bld, State &st, llvm::StringRef name) {
+        st.addAttribute("name", bld.getStringAttr(name));
+        auto reg = build_region(bld, st, {});
```
Comment (xlauko) [inline lib/vast/Dialect/HighLevel/HighLevelOps.cpp:207]: add `build_empty_region` which can be used on other places where `emplaceBlock` is called

**Review by xlauko — APPROVED** [review]

**Pull request merged**